Assignment1/assignment2/lib/utilities.dart

```diff
@@ -0,0 +1,112 @@
+import 'package:assignment2/users.dart';
+import 'dart:collection';
+
+Map<int, User> currentSession = Map();
+
+void nameHandler(String name) {
+  if (name.length == 0 || name.contains('[0-9]')) {
+    throw Exception(
+        "Name entered isn't valid, please enter appropriate details");
+  }
+}
+
+void ensurePositive(int number) {
+  if (number < 0) {
+    throw Exception(
+        "Number entered is not valid, please enter a positive number");
+  }
+}
+
+void ensureRollNumber(int rollNumber) {
+  if (currentSession.containsKey(rollNumber)) {
+    throw Exception(
+        "Roll Number already exists, please enter a new roll number");
+  }
+}
+
+List<String> acceptCourses(String? courses) {
+  List<String> registered = courses!.split(" ");
+  if (registered.length != 4) {
+    throw Exception("Registered courses should mandatorily be 4");
+  }
+  List<String> registered_courses = [];
+  registered.forEach((element) {
+    registered_courses.add(element.toUpperCase());
+  });
+  List<String> allowedCourses = ["A", "B", "C", "D", "E", "F"];
+  for (String element in registered_courses) {
+    if (allowedCourses.contains(element)) {
+      allowedCourses.remove(element);
+    } else {
+      throw Exception(
+          "Course Entered are wrong, please enter from the given courses");
+    }
+  }
+  return registered_courses;
+}
+
+void checkRollNumber(int rollNumber) {
+  if (!currentSession.containsKey(rollNumber)) {
+    throw Exception('Roll Number : $rollNumber does not exist in the database');
+  }
+}
+
+void loadCurrentSession(Map<int, User> map) {
+  currentSession = map;
+}
+
+SplayTreeMap<int, User> displayMap(String order, int option) {
+  int orderInt = 0;
+  if (order == 'a') {
+    orderInt = 1;
+  } else if (order == 'd') {
+    orderInt = -1;
+  } else {
+    throw Exception("Entered order for sorting is invalid");
+  }
+  SplayTreeMap<int, User> result = SplayTreeMap();
+  switch (option) {
+    case 1:
+      result = sortMapName(orderInt);
+      break;
+    case 2:
+      result = sortMapRollNumber(orderInt);
+      break;
+    case 3:
+      result = sortMapAge(orderInt);
+      break;
+    case 4:
+      result = sortMapAddress(orderInt);
+      break;
+    default:
+  }
+  return result;
+}
+
+SplayTreeMap<int, User> sortMapName(int order) {
+  Map<int, dynamic> map = currentSession;
+  SplayTreeMap<int, User> result = SplayTreeMap<int, User>.from(
+      map, (a, b) => map[a].name.compareTo(map[b].name) * order);
+  return result;
+}
+
+SplayTreeMap<int, User> sortMapRollNumber(int order) {
+  Map<int, dynamic> map = currentSession;
+  SplayTreeMap<int, User> result = SplayTreeMap<int, User>.from(map,
+      (a, b) => map[a].rollNumber > map[b].rollNumber ? order : -1 * order);
```

Comment:
Need a better explanation of what this statement does. Whenever you are writing some piece of code that might be a bit complex to understand for anyone it's better to add a comment above it explaining what we are doing and how is it done.

Reply:
done